src/appleseed/renderer/modeling/frame/frame.cpp

```diff
@@ -462,28 +490,462 @@ void Frame::denoise(
 
 namespace
 {
-    void create_parent_directories(const bf::path& file_path)
+
+    typedef vector<tuple<string, CanvasProperties, ImageAttributes>> CheckpointProperties;
+
+    //
+    // Interface used to save the rendering buffer in checkpoints.
+    //
+
+    class ShadingBufferCanvas
+      : public ICanvas
+    {
+      public:
+        ShadingBufferCanvas(
+            const Frame&                                frame,
+            PermanentShadingResultFrameBufferFactory*   buffer_factory)
+          : m_frame(frame)
+          , m_buffer_factory(buffer_factory)
+          , m_props(
+                m_frame.image().properties().m_canvas_width,
+                m_frame.image().properties().m_canvas_height,
+                m_frame.image().properties().m_tile_width,
+                m_frame.image().properties().m_tile_height,
+                (1 + m_frame.aov_images().size()) * 4 + 1,
```

Comment:
Where does that expression come from? Is it the responsability of `ShadingBufferCanvas` to compute that number of channels, are you trying to match the number of channels computed elsewhere?

Reply:
Why the additional `+ 1`?
Can you instead query the number of channels in the ShadingResultFrameBuffer instead?

Reply:
+1 Because i save the beauty image, just in case you want to correctly visualize the checkpoint